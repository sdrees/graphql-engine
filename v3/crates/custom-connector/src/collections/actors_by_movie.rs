use std::collections::BTreeMap;

use axum::{http::StatusCode, Json};
use ndc_models;

use crate::{
    query::Result,
    state::{AppState, Row},
    types::actor::get_actor_movie_id,
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "actors_by_movie".into(),
        description: Some("Actors parameterized by movie".into()),
        collection_type: "actor".into(),
        arguments: BTreeMap::from_iter([(
            "movie_id".into(),
            ndc_models::ArgumentInfo {
                argument_type: ndc_models::Type::Named { name: "Int".into() },
                description: None,
            },
        )]),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::new(),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<String, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    let movie_id = arguments.get("movie_id").ok_or((
        StatusCode::BAD_REQUEST,
        Json(ndc_models::ErrorResponse {
            message: "missing argument movie_id".into(),
            details: serde_json::Value::Null,
        }),
    ))?;
    let movie_id_int: i32 = movie_id
        .as_i64()
        .ok_or((
            StatusCode::BAD_REQUEST,
            Json(ndc_models::ErrorResponse {
                message: "movie_id must be a integer".into(),
                details: serde_json::Value::Null,
            }),
        ))?
        .try_into()
        .map_err(|_| {
            (
                StatusCode::BAD_REQUEST,
                Json(ndc_models::ErrorResponse {
                    message: "movie_id is out of range".into(),
                    details: serde_json::Value::Null,
                }),
            )
        })?;

    let mut actors_by_movie = vec![];

    for actor in state.actors.values() {
        let actor_movie_id_int = get_actor_movie_id(actor)?;
        if actor_movie_id_int == movie_id_int {
            actors_by_movie.push(actor.clone());
        }
    }

    Ok(actors_by_movie)
}
