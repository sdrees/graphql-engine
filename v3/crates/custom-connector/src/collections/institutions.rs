use std::collections::BTreeMap;

use ndc_models;

use crate::{
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "institutions".into(),
        description: Some("A collection of institutions".into()),
        collection_type: "institution".into(),
        arguments: BTreeMap::new(),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "InstitutionByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(state: &AppState) -> Result<Vec<Row>> {
    Ok(state.institutions.values().cloned().collect())
}
