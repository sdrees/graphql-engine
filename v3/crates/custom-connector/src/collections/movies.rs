use std::collections::BTreeMap;

use ndc_models;

use crate::{
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "movies".into(),
        description: Some("A collection of movies".into()),
        collection_type: "movie".into(),
        arguments: BTreeMap::new(),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "MovieByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(state: &AppState) -> Result<Vec<Row>> {
    Ok(state.movies.values().cloned().collect())
}
