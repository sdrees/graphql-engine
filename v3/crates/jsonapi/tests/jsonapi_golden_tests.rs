//! Tests that run JSONAPI to see if it works

use hasura_authn_core::{Identity, Role};
use reqwest::header::HeaderMap;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::str::FromStr;
use std::sync::Arc;

#[test]
fn test_get_succeeding_requests() {
    insta::glob!("passing/**/*.txt", |path| {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_all() // this enables time and IO
            .build()
            .unwrap();

        runtime.block_on(async {
            let TestEnvironment {
                jsonapi_catalog,
                resolved_metadata,
            } = test_environment_setup();

            let TestRequest { query, model_name } = test_request_setup(path);

            // always test in `default` subgraph for now
            let path = format!("/default/{model_name}");

            let http_context = execute::HttpContext {
                client: reqwest::Client::new(),
                ndc_response_size_limit: None,
            };

            let session = create_default_session();

            let result = jsonapi::handler_internal(
                Arc::new(HeaderMap::default()),
                Arc::new(http_context.clone()),
                Arc::new(session.clone()),
                &jsonapi_catalog,
                &resolved_metadata,
                axum::http::method::Method::GET,
                axum::http::uri::Uri::from_str(&path).unwrap(),
                query,
            )
            .await;

            match result {
                Ok(result) => {
                    insta::assert_debug_snapshot!(
                        format!("result_for_role_{}", session.role),
                        result
                    );
                }
                Err(e) => panic!("expected success, instead got {e}"),
            }
        });
    });
}

#[test]
fn test_get_failing_requests() {
    insta::glob!("failing/**/*.txt", |path| {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_all() // this enables time and IO
            .build()
            .unwrap();

        runtime.block_on(async {
            let TestEnvironment {
                jsonapi_catalog,
                resolved_metadata,
            } = test_environment_setup();

            let TestRequest { query, model_name } = test_request_setup(path);

            // always test in `default` subgraph for now
            let path = format!("/default/{model_name}");

            let http_context = execute::HttpContext {
                client: reqwest::Client::new(),
                ndc_response_size_limit: None,
            };

            let session = create_default_session();

            let result = jsonapi::handler_internal(
                Arc::new(HeaderMap::default()),
                Arc::new(http_context.clone()),
                Arc::new(session.clone()),
                &jsonapi_catalog,
                &resolved_metadata,
                axum::http::method::Method::GET,
                axum::http::uri::Uri::from_str(&path).unwrap(),
                query,
            )
            .await;

            insta::assert_debug_snapshot!(format!("error_for_role_{}", session.role), result);
        });
    });
}

#[test]
fn test_openapi_generation() {
    let TestEnvironment {
        jsonapi_catalog,
        resolved_metadata: _,
    } = test_environment_setup();

    for (role, state) in &jsonapi_catalog.state_per_role {
        let generated_openapi = jsonapi::openapi_schema(state);

        // if the test fails, let's take a look at what was generated
        dbg!(&serde_json::to_value(&generated_openapi)
            .unwrap()
            .to_string());

        insta::assert_json_snapshot!(
            format!("generated_openapi_for_role_{role}"),
            generated_openapi
        );
    }
}

struct TestRequest {
    query: jsonapi_library::query::Query,
    model_name: String,
}

struct TestEnvironment {
    jsonapi_catalog: jsonapi::Catalog,
    resolved_metadata: metadata_resolve::Metadata,
}

fn trim_newline(s: &mut String) {
    if s.ends_with('\n') {
        s.pop();
        if s.ends_with('\r') {
            s.pop();
        }
    }
}

fn test_environment_setup() -> TestEnvironment {
    // Setup test context
    let root_test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");
    let metadata_path = root_test_dir.join("static").join("metadata.json");

    let metadata_string = std::fs::read_to_string(metadata_path.clone()).unwrap_or_else(|error| {
        panic!("{}: Could not read file: {error}", metadata_path.display())
    });

    let metadata_json_value = serde_json::from_str(&metadata_string).unwrap();

    let input_metadata: open_dds::Metadata =
        open_dds::traits::OpenDd::deserialize(metadata_json_value, jsonpath::JSONPath::new())
            .unwrap();

    let configuration = get_metadata_resolve_configuration();

    let (resolved_metadata, _) = metadata_resolve::resolve(input_metadata, &configuration)
        .unwrap_or_else(|error| panic!("Could not resolve metadata: {error}",));

    let (jsonapi_catalog, _warnings) = jsonapi::Catalog::new(&resolved_metadata);

    TestEnvironment {
        jsonapi_catalog,
        resolved_metadata,
    }
}

fn test_request_setup(path: &Path) -> TestRequest {
    let directory = path.parent().unwrap();
    let model_name = path.file_stem().unwrap().to_str().unwrap();

    let mut query_params = std::fs::read_to_string(path).unwrap_or_else(|error| {
        panic!(
            "{}: Could not read file {path:?}: {error}",
            directory.display()
        )
    });

    // our input files contain trailing newlines that break the JSONAPI parser
    trim_newline(&mut query_params);

    let jsonapi_query = jsonapi_library::query::Query::from_params(&query_params);

    TestRequest {
        query: jsonapi_query,
        model_name: model_name.to_string(),
    }
}

// we will need to allow tests to define their own session options at some point
// the way that the GraphQL tests defined in `crates/engine/tests/execution.rs` do
fn create_default_session() -> hasura_authn_core::Session {
    //return an arbitrary identity with role emulation enabled
    let authorization = Identity::admin(Role::new("admin"));
    let role = Role::new("admin");
    let role_authorization = authorization.get_role_authorization(Some(&role)).unwrap();

    role_authorization.build_session(HashMap::new())
}

fn get_metadata_resolve_configuration() -> metadata_resolve::configuration::Configuration {
    let unstable_features = metadata_resolve::configuration::UnstableFeatures {
        enable_ndc_v02_support: false,
        enable_jsonapi: true,
        enable_aggregation_predicates: false,
    };

    metadata_resolve::configuration::Configuration { unstable_features }
}
