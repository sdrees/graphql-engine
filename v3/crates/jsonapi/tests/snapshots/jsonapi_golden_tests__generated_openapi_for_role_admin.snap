---
source: crates/jsonapi/tests/jsonapi_golden_tests.rs
expression: generated_openapi
---
{
  "openapi": "3.1.0",
  "info": {
    "title": "JSONAPI",
    "version": "0.1"
  },
  "paths": {
    "/default/Album": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Album]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "AlbumId",
                  "ArtistId",
                  "Title"
                ]
              }
            },
            "example": "AlbumId,ArtistId,Title"
          }
        ],
        "responses": null
      }
    },
    "/default/Articles": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Articles]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "article_id",
                  "title",
                  "author_id"
                ]
              }
            },
            "example": "article_id,title,author_id"
          }
        ],
        "responses": null
      }
    },
    "/default/Artist": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Artist]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "ArtistId",
                  "Name"
                ]
              }
            },
            "example": "ArtistId,Name"
          }
        ],
        "responses": null
      }
    },
    "/default/Authors": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Authors]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "author_id",
                  "first_name"
                ]
              }
            },
            "example": "author_id,first_name"
          }
        ],
        "responses": null
      }
    },
    "/default/Customer": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Customer]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "Address",
                  "City",
                  "Company",
                  "Country",
                  "CustomerId",
                  "Email",
                  "Fax",
                  "FirstName",
                  "LastName",
                  "Phone",
                  "PostalCode",
                  "State",
                  "SupportRepId"
                ]
              }
            },
            "example": "Address,City,Company,Country,CustomerId,Email,Fax,FirstName,LastName,Phone,PostalCode,State,SupportRepId"
          }
        ],
        "responses": null
      }
    },
    "/default/Employee": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Employee]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "Address",
                  "BirthDate",
                  "City",
                  "Country",
                  "Email",
                  "EmployeeId",
                  "Fax",
                  "FirstName",
                  "HireDate",
                  "LastName",
                  "Phone",
                  "PostalCode",
                  "ReportsTo",
                  "State",
                  "Title"
                ]
              }
            },
            "example": "Address,BirthDate,City,Country,Email,EmployeeId,Fax,FirstName,HireDate,LastName,Phone,PostalCode,ReportsTo,State,Title"
          }
        ],
        "responses": null
      }
    },
    "/default/Genre": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Genre]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "GenreId",
                  "Name"
                ]
              }
            },
            "example": "GenreId,Name"
          }
        ],
        "responses": null
      }
    },
    "/default/Invoice": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Invoice]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "BillingAddress",
                  "BillingCity",
                  "BillingCountry",
                  "BillingPostalCode",
                  "BillingState",
                  "CustomerId",
                  "InvoiceDate",
                  "InvoiceId",
                  "Total"
                ]
              }
            },
            "example": "BillingAddress,BillingCity,BillingCountry,BillingPostalCode,BillingState,CustomerId,InvoiceDate,InvoiceId,Total"
          }
        ],
        "responses": null
      }
    },
    "/default/InvoiceLine": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[InvoiceLine]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "InvoiceId",
                  "InvoiceLineId",
                  "Quantity",
                  "TrackId",
                  "UnitPrice"
                ]
              }
            },
            "example": "InvoiceId,InvoiceLineId,Quantity,TrackId,UnitPrice"
          }
        ],
        "responses": null
      }
    },
    "/default/MediaType": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[MediaType]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "MediaTypeId",
                  "Name"
                ]
              }
            },
            "example": "MediaTypeId,Name"
          }
        ],
        "responses": null
      }
    },
    "/default/Playlist": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Playlist]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "Name",
                  "PlaylistId"
                ]
              }
            },
            "example": "Name,PlaylistId"
          }
        ],
        "responses": null
      }
    },
    "/default/PlaylistTrack": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[PlaylistTrack]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "PlaylistId",
                  "TrackId"
                ]
              }
            },
            "example": "PlaylistId,TrackId"
          }
        ],
        "responses": null
      }
    },
    "/default/Track": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[Track]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "AlbumId",
                  "Bytes",
                  "Composer",
                  "GenreId",
                  "MediaTypeId",
                  "Milliseconds",
                  "Name",
                  "TrackId",
                  "UnitPrice"
                ]
              }
            },
            "example": "AlbumId,Bytes,Composer,GenreId,MediaTypeId,Milliseconds,Name,TrackId,UnitPrice"
          }
        ],
        "responses": null
      }
    },
    "/default/spatial_ref_sys": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[spatial_ref_sys]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "auth_name",
                  "auth_srid",
                  "proj4text",
                  "srid",
                  "srtext"
                ]
              }
            },
            "example": "auth_name,auth_srid,proj4text,srid,srtext"
          }
        ],
        "responses": null
      }
    },
    "/default/topology_layer": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[topology_layer]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "child_id",
                  "feature_column",
                  "feature_type",
                  "layer_id",
                  "level",
                  "schema_name",
                  "table_name",
                  "topology_id"
                ]
              }
            },
            "example": "child_id,feature_column,feature_type,layer_id,level,schema_name,table_name,topology_id"
          }
        ],
        "responses": null
      }
    },
    "/default/topology_topology": {
      "get": {
        "parameters": [
          {
            "name": "page[limit]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "5"
          },
          {
            "name": "page[offset]",
            "in": "query",
            "schema": {
              "type": "integer"
            },
            "example": "10"
          },
          {
            "name": "fields[topology_topology]",
            "in": "query",
            "schema": {
              "items": {
                "enum": [
                  "hasz",
                  "id",
                  "name",
                  "precision",
                  "srid"
                ]
              }
            },
            "example": "hasz,id,name,precision,srid"
          }
        ],
        "responses": null
      }
    }
  }
}
